client/asset/dcr/dcr.go

```diff
@@ -873,75 +886,87 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
 
 // FeeRate satisfies asset.FeeRater.
 func (dcr *ExchangeWallet) FeeRate() uint64 {
-	const confTarget = 2
+	const confTarget = 2 // 1 historically gives crazy rates
 	rate, err := dcr.feeRate(confTarget)
-	if err != nil {
+	if err != nil { // log and return 0
 		dcr.log.Errorf("feeRate error: %v", err)
 	}
 	return rate
 }
 
-// FeeRate returns the current optimal fee rate in atoms / byte.
+// feeRate returns the current optimal fee rate in atoms / byte.
 func (dcr *ExchangeWallet) feeRate(confTarget uint64) (uint64, error) {
-	// estimatesmartfee 1 returns extremely high rates on DCR.
-	if confTarget < 2 {
-		confTarget = 2
-	}
 	if feeEstimator, is := dcr.wallet.(FeeRateEstimator); is {
-		ratePerKB, err := feeEstimator.EstimateSmartFeeRate(dcr.ctx, int64(confTarget), chainjson.EstimateSmartFeeConservative)
-		if err == nil {
-			return convertFeeToUint(ratePerKB)
-		} else {
-			dcr.log.Errorf("Failed to get fee rate with estimate smart fee rate: %v", err)
+		dcrPerKB, err := feeEstimator.EstimateSmartFeeRate(dcr.ctx, int64(confTarget), chainjson.EstimateSmartFeeConservative)
+		if err == nil && dcrPerKB > 0 {
+			return dcrPerKBToAtomsPerByte(dcrPerKB)
+		}
+		if err != nil {
+			dcr.log.Warnf("Failed to get local fee rate estimate: %v", err)
+		} else { // dcrPerKB == 0
+			dcr.log.Warnf("Local fee estimate is zero.")
 		}
 	}
+
 	// Either SPV wallet or EstimateSmartFeeRate failed.
 	if !dcr.apiFeeFallback {
-		return 0, fmt.Errorf("fee rate estimation unavailable")
+		return 0, fmt.Errorf("fee rate estimation unavailable and external API is disabled")
 	}
-	dcr.log.Debug("Retrieving fee rate from external API")
-	ratePerKB, err := externalFeeEstimator(dcr.ctx, dcr.network, confTarget)
+
+	now := time.Now()
+
+	dcr.oracleFeesMtx.Lock()
+	defer dcr.oracleFeesMtx.Unlock()
+	oracleFee := dcr.oracleFees[confTarget]
+	if now.Sub(oracleFee.stamp) < freshFeeAge {
+		return oracleFee.rate, nil
+	}
+
+	dcr.log.Debugf("Retrieving fee rate from external fee oracle for %d target blocks", confTarget)
+	dcrPerKB, err := fetchFeeFromOracle(dcr.ctx, dcr.network, confTarget)
```

Comment:
This sequence is a little scary. Consider the case when the oracle is somehow non-responsive. The first rate caller is stuck here for 4 seconds. Any other callers within that 4 seconds will line up. When the timeout occurs, subsequent requesters will still wait in line to issue requests themselves, each taking 4 seconds in turn.

I think we need an error mode, so that subsequent requests are blocked for some time after an error or two. On the `FeeRate` route, the client may still be able to get a fee suggestion from the DEX server.

Reply:
> When the timeout occurs, subsequent requesters will still wait in line to issue requests themselves, each taking 4 seconds in turn

Good call.  That's an ugly outcome.
I think you're right that there needs to be a "failing" state that expires after a bit.